### PR TITLE
Fixing overlay network configuration failure.

### PR DESCRIPTION
**Pull request by asridharan**
- Added a new state: `STATUS_CONFIGURING` to identify when an overlay network is in the middle of a configuration.
- Modified continuation `__configure` to capture and failures in `iptables` provisioning.
- Modified the handling of register acknowledgement from the master. The agent will now drop any register acknowledgment requests if there is a failure to configure any of the overlay network. This would force the agent to re-register with the master, triggering a re-configuration of the overlay network. Also, added a max number of mis-configurations that the agent module can tolerate before bailing out.

### CODE REVIEW

File: src/agent/manager.cpp [hunk 1]
```diff
@@ -303,19 +316,9 @@ class ManagerProcess : public ProtobufProcess<ManagerProcess>
...
-        // we should simply wait for the new overlays to be configured
-        // and that will trigger another '_updateAgentOverlays()'.
-        LOG(INFO) << "Ignored sending registered message because overlay "
-                  << "network '" << overlay.info().name()
-                  << "' is still being configured";
-        return;
-      }
+      // Every overlay network should have a state, and it should be
+      // either in `STATUS_OK` or `STATUS_FAILED`.
+      CHECK(overlay.has_state());
```
Comment (asridharan) [inline src/agent/manager.cpp:321]: Add explicit check for `STATUS_OK` and STATUS_FAILED`

File: src/agent/manager.cpp [hunk 2]
```diff
@@ -476,13 +499,16 @@ class ManagerProcess : public ProtobufProcess<ManagerProcess>
...
-    // exist insert one in the POSTROUTING change of the NAT table.
-    // The below command is a chain of three commands:
+    // exist insert one in the POSTROUTING chain of the NAT table.
+    // The below command is a script consisting of three commands:
     // <set ipset> && <check iptables rule exists> ||
     // <insert iptables rule>
     if (!overlays[name].has_mesos_bridge() && 
         !overlays[name].has_docker_bridge()) {
-      return __configure(name, string(""));
+      overlays[name].mutable_state()->set_status(
```
Comment (asridharan) [inline src/agent/manager.cpp]: Revert this change.

File: src/agent/manager.cpp [hunk 3]
```diff
@@ -508,12 +534,21 @@ class ManagerProcess : public ProtobufProcess<ManagerProcess>
       LOG(INFO) << "Insert following iptables rule for overlay " << name
                 << ": " << command.get();
 
+      // We are using `await` here so that we can capture the failure
+      // case of the iptables script. Without using `await` we would
+      // have to use `onAny` which would break the continuation of
+      // this callback.
       return runScriptCommand(command.get())
         .then(defer(
               self(),
```
Comment (asridharan) [inline src/agent/manager.cpp]: Just add a place holder lambda.

File: src/agent/manager.cpp [hunk 4]
```diff
@@ -227,20 +231,31 @@ class ManagerProcess : public ProtobufProcess<ManagerProcess>
...
+      // if it is in `STATUS_CONFIGURING` state, we will skip the
+      // configuration. This might change soon.
+      if (overlays.contains(name) &&
+          overlays.at(name).has_state() &&
+          overlays.at(name).state().has_status() ) {
+        OverlayState::Status status = overlays.at(name).state().status();
+
+        if ((status == OverlayState::STATUS_OK) ||
+            (status == OverlayState::STATUS_CONFIGURING)) {
+          LOG(INFO)
```
Comment (jieyu) [inline src/agent/manager.cpp]: Style nits:

```
LOG(INFO) << ...
          << ...
```

File: src/agent/manager.cpp [hunk 5]
```diff
@@ -227,20 +231,31 @@ class ManagerProcess : public ProtobufProcess<ManagerProcess>
...
+            << "Skipping configuration for overlay network '"
+            << name << "' as it "
+            << ((status == OverlayState::STATUS_OK) ?
+                "has been " : "is being ")
+            <<"configured.";
```
Comment (jieyu) [inline src/agent/manager.cpp]: space after `<<`
